components/service/glean/src/main/java/mozilla/components/service/glean/storages/StringListsStorageEngine.kt

```diff
@@ -0,0 +1,87 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.annotation.SuppressLint
+import mozilla.components.service.glean.CommonMetricData
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This singleton handles the in-memory storage logic for string lists. It is meant to be used by
+ * the Specific String List API and the ping assembling objects. No validation on the stored data
+ * is performed at this point: validation must be performed by the Specific String List API.
+ *
+ * This class contains a reference to the Android application Context. While the IDE warns
+ * us that this could leak, the application context lives as long as the application and this
+ * object. For this reason, we should be safe to suppress the IDE warning.
+ */
+@SuppressLint("StaticFieldLeak")
+internal object StringListsStorageEngine : StringListsStorageEngineImplementation()
+
+internal open class StringListsStorageEngineImplementation(
+    override val logger: Logger = Logger("glean/StringsStorageEngine")
+) : GenericScalarStorageEngine<List<String>?>() {
+
+    companion object {
+        // Maximum length of any list
+        const val MAX_LIST_LENGTH_VALUE = 20
+    }
+
+    @SuppressWarnings("unchecked")
+    override fun deserializeSingleMetric(value: Any?): List<String>? {
+        return value as? List<String>
+    }
+
+    /**
+     * Record a string list in the desired stores.
+     *
+     * @param metricData object with metric settings
+     * @param value the string list value to add
+     */
+    @Synchronized
+    fun add(
+        metricData: CommonMetricData,
+        value: List<String>
+    ) {
+        // Use a custom combiner to add the string to the existing list rather than overwriting
+        super.recordScalar(metricData, value) { currentValue, newValue ->
+            currentValue?.let {
+                return@let if (it.count() + value.count() <= MAX_LIST_LENGTH_VALUE) {
+                    it + newValue!!
+                } else {
+                    // Log a warning and truncate the newValue list to fill in to the maximum
+                    logger.warn("${metricData.category}.${metricData.name} - " +
+                            "discarding extra values. List limit $MAX_LIST_LENGTH_VALUE reached.")
```

Comment:
nit: let's properly indent "discarding..." so that it aligns with `"${metricData..."` from the previous line

Reply:
Will ktlint/detect ding me for the extra whitespace?

Reply:
Good questions :) I don't know! But you can run klint / detekt locally and check this out ;) (both from the IDE and the command line)

Reply:
I have this in my `.git/hooks/pre-commit` file for the project and I've found it useful:

```
#!/bin/sh
./gradlew --no-daemon ktlint detekt
```